libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -698,8 +698,11 @@ string ABIFunctions::abiEncodingFunctionCompactStorageArray(
 					switch and(slotValue, 1)
 					case 0 {
 						// short byte array
-						mstore(pos, and(slotValue, not(0xff)))
-						ret := add(pos, <lengthPaddedShort>)
+						ret := pos //assume empty array
+						if gt(length, 0) {
+							mstore(pos, and(slotValue, not(0xff)))
+							ret := add(pos, <lengthPaddedShort>)
+						}
```

Comment:
Following @chriseth's suggestion in https://github.com/ethereum/solidity/issues/12880#issuecomment-1122292315, this should look like this:
```suggestion
						mstore(pos, and(slotValue, not(0xff)))
						ret := add(pos, mul(<lengthPaddedShort>, iszero(length)))
```

Your version requires a jump. `JUMPI` opcode is 10 gas while `MUL` is just 5.

But it would be best to measure it to be sure (the final cost difference might be different due to the optimizer). You might also be able to come up with something even cheaper.

Reply:
You can compare costs of different versions by using `isoltest` to update gas expectations in sematic tests:
```bash
isoltest --accept-updates --enforce-gas-cost --optimize
```
When you run this, isoltest will update these tests in your working copy and you'll be able to see what the gas difference is. The gas update for all tests takes some time so for quick testing you might want to select just one specific test. You can do that via the `--test` option. E.g. `--test=semanticTests/various/skip_dynamic_types_for_structs`.

Also, when you're done, you'll have to update these expectations and commit them for all soltest runs in CI to pass. To update them all, you need to run the above `isoltest` command twice - once with and once without `--optimize`. For example right now the `t_ubu_soltest_all` job is failing in your PR exactly for this reason:

```xml
<testcase assertions="4" classname="semanticTests.various" name="skip_dynamic_types_for_structs" time="0.358608">
<failure message="failure" type="assertion error"><![CDATA[
ASSERTION FAILURE:
- file   : boostTest.cpp
- line   : 111
- message: Test expectation mismatch.
Running via Yul:
Expected result:
// g() -> 2, 6
// gas irOptimized: 178805
// gas legacy: 180753
// gas legacyOptimized: 179472

Obtained result:
// g() -> 2, 6
// gas irOptimized: 178868 [+63 (+0%)]
// gas legacy: 180753
// gas legacyOptimized: 179472
```

By the way, something seems broken since [CircleCI does not show the failing test output](https://app.circleci.com/pipelines/github/ethereum/solidity/24096/workflows/6e82f6f8-585f-410c-b863-7e5147a86cf2/jobs/1058493/tests). We'll need to fix that but for now I was able to find the failing test by looking at the artifacts directly: